amaranth/vendor/gowin.py

```diff
@@ -124,17 +123,19 @@ def _osc_props(self):
         else:
             assert False
 
-        if ((not isinstance(self.osc_freq, int))
-            and (not isinstance(self.osc_freq, float))):
-                raise ValueError("osc_freq must be an int or float")
+        if not hasattr(self, "osc_frequency"):
+            raise AttributeError('default_clk "OSC" requires attribute "osc_frequency" to be set.')
 
-        # osc_freq is given in MHz
-        div = int(props["freq"] / (self.osc_freq * 1e6))
+        if not isinstance(self.osc_frequency, int):
+            raise ValueError("osc_frequency must be an int.")
```

Comment:
This would be a `TypeError`, but in any case we don't really need to enforce it here.

Reply:
Ok.